### DISCUSSION TITLE
tutorials/export/exporting_for_ios

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#159] [User-contributed notes]
# tutorials/export/exporting_for_ios

These are the steps to load a Godot project in Xcode. This allows you to build and deploy to an iOS device, build a release for the App Store, and do everything else you can normally do with Xcode....

https://docs.godotengine.org/en/stable/tutorials/export/exporting_for_ios.html

**Comment by etherealxx** [comment]
Can anyone else confirm the Active Development (linking Godot project folder to Xcode) is broken/not working for xcode 16.2?
Are there any open issue of it currently?

**Reply by yemibox51** [reply]
Yes, here: https://github.com/godotengine/godot/issues/98474

**Comment by alkhimey** [comment]
"Steps to link a Godot project folder to Xcode" does not work for me. (Godot 4.3, Xcode 16.0)
There are several issues. First, in the example the "Add to target" is selected, which causes many linker issues as the project folder (.godot) contains libraries which will be automatically added to linked files.
I solved this by first removing .godot folder, adding the main folder to Xcode project and then reading the .godot folder back, so there is a reference but it is not copied to the target. This will still add some "junk" to the "copy bundle resources". It is ok to remove all of them except (if I understand correctly how this scheme supposed to work) a PCK file.
The other issue is that "godot_path" should be the path to a folder, so the value need to be set to "." (iOS bundle is flat). However  after deleting the pck file generated during the inital export, building with Xcode does not produce new pck file, it is not copied to the bundle and thus I get the error:

Error: Couldn't load project data at path ".". Is the .pck file missing?
If you've renamed the executable, the associated .pck file should also be renamed to match the executable's name (without the extension).

**Reply by yemibox51** [reply]
Issue has been reported: https://github.com/godotengine/godot/issues/98474. Some users have found a workaround.

**Comment by brylie** [comment]
> You can find the code over at developer.apple.com next to your name in the top right corner.

This seems to no longer be true. On developer.apple.com/account, I only see my name in the upper right corner. The docs may need an update to describe how to get the App Store Team ID.

**Reply by bruvzg** [reply]
https://developer.apple.com/help/glossary/team-id/ - it should be in the “Membership details.”

It's also possible to get Team ID directly from a certificate in your Keychain, it's called “Organizational Unit” in the certificate details.

**Reply by alkhimey** [reply]
And from the terminal:

```security find-identity -v -p codesigning```

**Reply by bruvzg** [reply]
> security find-identity -v -p codesigning

This will only print User ID, but not the Team ID, these are different.

**Reply by alkhimey** [reply]
What is User ID? Never heard about it, but every opportunity is good for learning :) 
For me the above command prints my Team ID. It prints the name of the certificates, and the docs say it contains the Team ID

[See footnote here.](https://developer.apple.com/help/account/certificates/certificates-overview/)
> Note: In your keychain, a signing certificate name contains a hint to the type, and includes the team name and Team ID.

I should have mentioned that distribution certificate should be used.

**Reply by bruvzg** [reply]
Indeed, it seems to be different only for Development certificates, and the same for Distribution, so this can be used as well.

**Reply by alkhimey** [reply]
Just for the sake of perfectionism, my friend copilot, recommends the following command that will print the organizational unit (one or more, but most people that have more than one, know what they are doing and don't need this shortcut anyway):

```
security find-identity -v -p codesigning | awk -F '"|:' '/"/ {print $2}' | while read common_name; do
  security find-certificate -c "$common_name" -p | openssl x509 -noout -subject | grep -o "OU=[^,]*" | cut -d= -f2
done | sort -u

```

**Comment by gtodd876** [comment]
I have a PR started to fix the outdated iOS information: https://github.com/godotengine/godot-docs/pull/10900

**Comment by Serimert90** [comment]
Export fails for me

error: Communication with Apple failed: Your team has no devices from which to generate a provisioning profile. Connect a device to use or manually add device IDs in Certificates, Identifiers & Profiles. https://developer.apple.com/account/

**Comment by anilkonac** [comment]
To get the debugger (i.e., pausing at breakpoints) working after using the Remote Deploy button, make sure to change the Editor Settings → Network → Debug → Remote Host value to your PC’s local IP address (instead of 127.0.0.1).

After several hours of searching online for an iOS-specific solution without success, I eventually found the fix in a thread about Android debugging: https://forum.godotengine.org/t/remote-debugger-on-android-device-issue/84642/2

**Comment by ESnider** [comment]
Can you please add a paragraph or sentence about iOS Simulator support? I think it's totally been pulled from Godot, but would be great to have that 100% confirmed here so developers don't need to wonder about it. Also - if that's the case it would be helpful to have a very brief explanation of connecting an iOS device, choosing it from the pop-up etc since almost everything else is here?

**Reply by Calinou** [reply]
The iOS simulator isn't really useful on Apple Silicon Macs, as you can run iOS apps directly. It's still supported on Intel Macs though, as they actually need the iOS simulator to be able to run iOS apps.

**Reply by ESnider** [reply]
Oh, wow – I'm embarrassed to say I didn't realize Xcode was set up to run iOS apps directly. I've been making iOS games for a really long time and I guess was so used to the Simulator I didn't know! I've got a new M4 Mac Mini and it's pretty good times now. Thanks!

**Reply by gkwaerp** [reply]
Unless I'm misunderstanding something, I would disagree with Calinou here. Being able to use the iOS simulator would be very beneficial in some circumstances. For example, I'm trying to track down why our game crashes during load on some older devices.

(Note that this is not related to renderer used. The game starts, but then crashes midway through loading a scene. Right now our main suspect is memory related issues, but without access to any hardware that old and no iOS simulator, we're just testing things blindly.)

That said, if Simulator is intentionally not supported, it might still be worthwhile having a sentence about in the documentation, to prevent more people spending ages and ages trying to make it work.

**Reply by Calinou** [reply]
The iOS simulator doesn't support enough GPU features to run Godot's Forward+/Mobile renderers:

- https://github.com/godotengine/godot/issues/74227

Also, its GPU emulation is not 1:1 to any physical iOS device out there, so it's quite likely to have bugs that only occur in the iOS simulator.

That said, it appears the iOS simulator isn't supported at all since Godot 4.4 anyway, even for Compatibility:

- https://github.com/godotengine/godot/issues/102149

It's still open as an issue that could be fixed, but Forward+/Mobile support can't happen unless support for image cube arrays is somehow made optional. https://github.com/godotengine/godot/pull/108127 and/or https://github.com/godotengine/godot/pull/107902 might make this optional but I'm not 100% sure.
